src/core/mount.c

```diff
@@ -142,6 +152,104 @@ static bool mount_needs_quota(const MountParameters *p) {
                                  "usrquota\0" "grpquota\0" "quota\0" "usrjquota\0" "grpjquota\0");
 }
 
+static int alloc_mount_hash(void) {
+        Hashmap *by_mnt_id;
+        Hashmap *by_device;
+
+        by_mnt_id = hashmap_new(NULL);
+        if (!by_mnt_id)
+                return -ENOMEM;
+        by_device = hashmap_new(NULL);
+        if (!by_device) {
+                hashmap_free(by_mnt_id);
+                return -ENOMEM;
+        }
+        mount_hash.by_mnt_id = by_mnt_id;
+        mount_hash.by_device = by_device;
+
+        return 0;
+}
+
+static int mount_hash_put(Mount *m, int mnt_id, const char *what) {
+        Mount *mnt;
+        int dev_ref;
+        int r;
+
+        assert(m);
+
+        if (!mount_hash.by_mnt_id) {
+                if (alloc_mount_hash() != 0)
+                        return -ENOMEM;
+        }
+
+        if (mnt_id == -1)
+                return 0;
+
+        mnt = hashmap_get(mount_hash.by_mnt_id, INT_TO_PTR(mnt_id));
+        if (!mnt) {
+                r = hashmap_put(mount_hash.by_mnt_id, INT_TO_PTR(mnt_id), m);
+                if (r < 0)
+                        return -ENOMEM;
+        }
+
+        dev_ref = PTR_TO_INT(hashmap_get(mount_hash.by_device, what));
+        if (dev_ref == 0)
+                dev_ref = 1;
+        else
+                dev_ref++;
+        r = hashmap_replace(mount_hash.by_device, what, INT_TO_PTR(dev_ref));
+        if (r < 0)
+                return -ENOMEM;
+
+        return 0;
+}
+
+static int mount_hash_remove(Mount *m) {
+        MountParameters *p;
+        Mount *mnt;
+        int dev_ref;
+        int r;
+
+        assert(m);
+
+        if (!mount_hash.by_mnt_id)
+                return 0;
+
+        p = &m->parameters_proc_self_mountinfo;
+
+        mnt = hashmap_get(mount_hash.by_mnt_id, INT_TO_PTR(p->mnt_id));
+        if (mnt)
+                (void) hashmap_remove(mount_hash.by_mnt_id, INT_TO_PTR(p->mnt_id));
+
+        r = 0;
+        dev_ref = PTR_TO_INT(hashmap_get(mount_hash.by_device, p->what));
+        if (dev_ref <= 1)
+                (void) hashmap_remove(mount_hash.by_device, p->what);
```

Comment:
should probably be enclosed in `assert_se()` since this cannot possibly fail unless it's a programming error. i.e.

```c
assert_se(hashmap_remove(mount_hash.by_device, p->what) == INT_TO_PTR(dev_ref));
```

(background: assert_se() is like assert(), but may have side-effects, and is never optimized away, even if NDEBUG is set, unlike assert().)

Reply:
These comments about the hashmap implementation are really useful, I did look but it's a little complex, thanks for these.